### ISSUE TITLE
Software serial and PWM functions?

### ISSUE THREAD

**Issue by xiahualiu** [raspberrypi/pico-sdk#555]
So I have not programmed the Pico yet, but I have a few questions about the SDK first, I can be wrong so you can correct me if I am wrong.

I saw the SDK has hardware UART and pwm APIs, but I did not find the software UART and PWM APIs.

Is there any way to do software serial and pwm communication using the GPIOs? This should not be a problem on the hardware.

The reason is because I have 5 sensors, they all have serial interfaces. Unlike SPI and I2C, serial and pwm is 1-to-1 so I need 5 serial ports.

Thanks!

**Comment by Wren6991** [comment]
> Is there any way to do software serial and pwm communication over GPIOs?

Yes we have something called PIO. You can do 8 UART TXs on different pins using PIO for example.

There is a chapter on this in the C/C++ SDK PDF, which is a good place to start:

![image](https://user-images.githubusercontent.com/1298595/132187286-ac62c1d4-705f-4496-846a-42f2f8441ed0.png)

Hardware details are in the datasheet:

![image](https://user-images.githubusercontent.com/1298595/132187373-914181e9-08d4-4185-a3d6-a0a0f249e04b.png)

Software example for PWM using PIO:

https://github.com/raspberrypi/pico-examples/tree/master/pio/pwm

Software example for serial TX using PIO:

https://github.com/raspberrypi/pico-examples/tree/master/pio/uart_tx

You can of bitbang a UART or PWM using a GPIO pin if you want to. This works the same way as it does on other microcontrollers. It's not a very good approach -- PIO works out better because the timing is completely predictable, and it doesn't require any CPU cycles.

Hope this helps.
